### PR TITLE
Support for resource limits for containers (Alpha Feature)

### LINKED ISSUE #3595: Expose or utilize docker's rlimit support
https://github.com/docker/docker/issues/4717#issuecomment-70373299

Now that this is in, we should define how we want to use it.

### PR DESCRIPTION
**Pull request by dims**
The idea here is that folks should not have to use full privileged containers to be able to tweak small things like maximum file descriptors ("nofile") or maximum number of bytes of memory that may be locked into RAM ("memlock"). Today they write extra scripts that get baked into their containers (see example - https://github.com/kubernetes/kubernetes/issues/3595#issuecomment-354481790). Which is extra hassle and unnecessary as docker already supports specifying ulimits (see https://docs.docker.com/engine/reference/commandline/run/#set-ulimits-in-container-ulimit). 

So let use use the Requests/Limits mechanism to let the end users specify what they need in their container definitions.  Right now since docker supports passing these directly let us experiment this capability on just docker. 

Fixes #3595.

```release-note
Alpha feature to support specifying Resource Limits. Works only on docker runtime.
```

### CODE REVIEW

**Comment by dims** [comment]
Looks like the valid values are:
https://github.com/docker/go-units/blob/master/ulimit.go#L48-L62

the man page with details on the values is here:
http://man7.org/linux/man-pages/man2/getrlimit.2.html

**Comment by dims** [comment]
Questions : 
- How do we support other CRI implementations?
- Do we support all the rlimit options? which ones other than `nofile` and `memlock` makes sense?

**Comment by dims** [comment]
cc @liggitt

**Comment by adohe** [comment]
@dims would like to see an example pod config to show how we use.

**Comment by adohe** [comment]
```
Do we support all the rlimit options? which ones other than nofile and memlock makes sense?
```
I am afraid not all, we need to figure out how this rlimit options coordinate with Kube cgroup based resource isolation.

**Comment by dims** [comment]
@adohe it's just requests and limits with the `rlimit-` prefix. requests end up being the `Soft` value and the limit is the `Hard` value

```
containers:
- image: nginx
  imagePullPolicy: Always
  name: default-mem-demo-ctr
  resources:
    limits:
      rlimit-nofile: 2048
    requests:
      rlimit-nofile: 1024
```

**Comment by adohe** [comment]
@dims I would think whether we should regard this as `standard resource` or `extended resource`, also I think we should make this as `extended resource` which means its unknown to the system.

**Comment by adohe** [comment]
if regard it as `extended resource`, we should use fully qualified name in pod spec.

File: pkg/apis/core/helper/helpers.go [hunk 1]
```diff
@@ -37,6 +37,10 @@ func IsHugePageResourceName(name core.ResourceName) bool {
 	return strings.HasPrefix(string(name), core.ResourceHugePagesPrefix)
 }
 
+func IsRlimitResourceName(name string) bool {
+	return strings.HasPrefix(string(name), "rlimit-")
```
Comment (adohe) [inline pkg/apis/core/helper/helpers.go]: how about use prefix `ulimit-`, I think this could be more meaningful as we are support docker -ulimit.

**Comment by dims** [comment]
@adohe : adding it as a extended resource would be too much ( looking at how we have to advertise it etc - https://kubernetes.io/docs/tasks/administer-cluster/extended-resource-node/ ). Since the ulimits is per process as well (not per node) it would be disconnect too. So i would prefer it to be a `standard resource`

File: pkg/apis/core/helper/helpers.go (hunk 1 above)
Comment (dims) [reply pkg/apis/core/helper/helpers.go]: yep, i can do this quickly

**Comment by ConnorDoyle** [comment]
+1 to @dims point. If you want to enforce isolation of an ER, the only supported way right now is to create a device plugin. The API does have per-container semantics, but unfortunately the device plugin interface doesn’t allow setting arbitrary container resources. You’d have to extend the API there to enforce ulimit via a device plugin. Furthermore, device plugin feels like the wrong abstraction for something like ulimit.

**Comment by adohe** [comment]
after a discussion with @ConnorDoyle we both agree that ER is not the right way to support `ulimit`, I think we could support `ulimit` as standard resource, just put it forward.

**Comment by dims** [comment]
@kubernetes/sig-node-pr-reviews

File: pkg/kubelet/kubelet.go [hunk 2]
```diff
@@ -890,6 +890,11 @@ func NewMainKubelet(kubeCfg *kubeletconfiginternal.KubeletConfiguration,
 			glog.Errorf("Accelerators feature is supported with docker runtime only. Disabling this feature internally.")
 		}
 	}
+	if utilfeature.DefaultFeatureGate.Enabled(features.SupportUlimits) {
+		if containerRuntime != kubetypes.DockerContainerRuntime {
+			glog.Warning("ULimits support is supported with docker runtime only. UFLimits specified will not have any effect when using %v.", containerRuntime)
```
Comment (tallclair) [inline pkg/kubelet/kubelet.go]: s/Warning/Warningf/

File: pkg/kubelet/kuberuntime/kuberuntime_container.go [hunk 3]
```diff
@@ -233,6 +235,34 @@ func (m *kubeGenericRuntimeManager) generateLinuxContainerConfig(container *v1.C
 		SecurityContext: m.determineEffectiveSecurityContext(pod, container, uid, username),
 	}
 
+	if utilfeature.DefaultFeatureGate.Enabled(features.SupportUlimits) {
+		lc.Resources.Ulimits = make(map[string]*runtimeapi.LimitValue)
+		for key, value := range container.Resources.Requests {
+			if strings.HasPrefix(string(key), "ulimit-") {
```
Comment (tallclair) [inline pkg/kubelet/kuberuntime/kuberuntime_container.go]: How about using a prefix like `ulimits.k8s.io/`? I think it's more consisctent with other resources.

File: pkg/features/kube_features.go [hunk 4]
```diff
@@ -238,6 +238,12 @@ const (
 	// Mount secret, configMap, downwardAPI and projected volumes ReadOnly. Note: this feature
 	// gate is present only for backward compatability, it will be removed in the 1.11 release.
 	ReadOnlyAPIDataVolumes utilfeature.Feature = "ReadOnlyAPIDataVolumes"
+
+	// owner: @dims
+	// alpha: v1.10
```
Comment (tallclair) [inline pkg/features/kube_features.go]: 1.11? Should add this to the milestone if that's the goal.

File: pkg/features/kube_features.go [hunk 5]
```diff
@@ -238,6 +238,12 @@ const (
...
+	//
+	// Implement support for ulimits in containers
+	SupportUlimits utilfeature.Feature = "SupportUlimits"
```
Comment (tallclair) [inline pkg/features/kube_features.go]: You should add this to `defaultKubernetesFeatureGates` below.

File: pkg/kubelet/apis/cri/runtime/v1alpha2/api.proto [hunk 6]
```diff
@@ -482,6 +482,11 @@ message KeyValue {
     string value = 2;
 }
 
+message LimitValue {
```
Comment (tallclair) [inline pkg/kubelet/apis/cri/runtime/v1alpha2/api.proto]: nit: Add comments

**Review by tallclair — COMMENTED** [review]
/cc @dashpole @yujuhong

File: pkg/features/kube_features.go (hunk 4 above)
Comment (dims) [reply pkg/features/kube_features.go]: Done

File: pkg/features/kube_features.go (hunk 5 above)
Comment (dims) [reply pkg/features/kube_features.go]: ack. will do

File: pkg/kubelet/kuberuntime/kuberuntime_container.go (hunk 3 above)
Comment (dims) [reply pkg/kubelet/kuberuntime/kuberuntime_container.go]: i tried that in a previous iteration and ran into trouble. will dig into my notes and post back

**Comment by dims** [comment]
@tallclair @dashpole @yujuhong can we please get this in early in the cycle to get some feedback and tweak things if necessary.

File: pkg/apis/core/helper/helpers.go [hunk 7]
```diff
@@ -246,7 +251,7 @@ var standardResources = sets.NewString(
 
 // IsStandardResourceName returns true if the resource is known to the system
 func IsStandardResourceName(str string) bool {
-	return standardResources.Has(str) || IsQuotaHugePageResourceName(core.ResourceName(str))
+	return standardResources.Has(str) || IsQuotaHugePageResourceName(core.ResourceName(str)) || IsUlimitResourceName(str)
```
Comment (liggitt) [inline pkg/apis/core/helper/helpers.go]: are we adding more unqualified resource names? I thought the direction was pushing the other way (qualifying cpu, memory, etc, with `kubernetes.io/` prefixes)

Comment (liggitt) [reply pkg/apis/core/helper/helpers.go]: (maybe I'm misreading https://github.com/kubernetes/kubernetes/issues/50658#issuecomment-376983173)

Comment (liggitt) [reply pkg/apis/core/helper/helpers.go]: in any case, is this allowing setting unqualified resources in pod specs we didn't previously allow, as part of an alpha feature? if so, should it only be allowed when the feature is enabled? what are the implications if you set this and the backing runtime doesn't support it?

Comment (dims) [reply pkg/apis/core/helper/helpers.go]: @liggitt we are warning when the underlying runtime does not support it - https://github.com/kubernetes/kubernetes/pull/58904/commits/65e7d63e371e2d2aaedb2da32042d92509e077d2#diff-bf28da68f62a8df6e99e447c4351122dR913

we could elevate that to a fatal error?

Also, do you want me to add a check for `utilfeature.DefaultFeatureGate.Enabled(features.SupportUlimits)` here?

Comment (liggitt) [reply pkg/apis/core/helper/helpers.go]: > we are warning when the underlying runtime does not support it - 65e7d63#diff-bf28da68f62a8df6e99e447c4351122dR913

I don't have a good sense for the way unknown/unsupported resources have been handled in the past. Intuitively, if I declare my pod requires resource x and it is unsupported, I would expect my pod to not be run, and the status reflect that the resource I required wasn't supported.

> do you want me to add a check for utilfeature.DefaultFeatureGate.Enabled(features.SupportUlimits) here?

if we're calling this an alpha feature, and validation prevented pods from requesting this in the past, we should prevent the possibility of persisted pods with this resource until we're confident enough in the feature's stability/longevity to promote it

File: pkg/kubelet/kubelet.go (hunk 2 above)
Comment (dims) [reply pkg/kubelet/kubelet.go]: Done

File: pkg/kubelet/apis/cri/runtime/v1alpha2/api.proto (hunk 6 above)
Comment (dims) [reply pkg/kubelet/apis/cri/runtime/v1alpha2/api.proto]: Done

File: pkg/apis/core/helper/helpers.go (hunk 7 above)
Comment (dims) [reply pkg/apis/core/helper/helpers.go]: ack, reworking so if the resource is needed and the feature is absent then we error out.

**Comment by vishh** [comment]
Has ulimits evolved beyond the issues posted in https://github.com/kubernetes/kubernetes/issues/3595? The main problem then was that ulimits do not have any correlation with cgroups and hence containers.
Also this PR feels like an idea that I'd recommend discussing further in https://github.com/kubernetes/kubernetes/issues/3595 before reviewing a PR for it.

**Comment by dims** [comment]
@vishh there is a discussion there (see starting Jan 9th).

**Comment by vishh** [comment]
@dims that discussion does not seem to provide a full coverage on the ulimits that are available and their scope within a container. I'd love to see a table that shows the list of ulimits, their scope (node vs container vs process) and their interaction with cgroups and namespaces.
This feels similar to the sysctl design discussions where we chose to include a set of safe sysctls into the core and let the rest be handled out of band.

**Comment by dims** [comment]
@bgrant0607 by `container spec` do you mean https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/api/core/v1/types.go#L1920 ?

**Comment by bgrant0607** [comment]
@dims Yes, a sibling to resources and to security context.

**Comment by adohe** [comment]
@bgrant0607 but I do think `ulimits` should be regard as `resources`, I see docker put ulimits configuration in the `Resources` struct, as you can see [here](https://github.com/moby/moby/blob/master/api/types/container/host_config.go#L337)

**Comment by bgrant0607** [comment]
@adohe There are many things in that list that wouldn't make sense as scheduled finite countable resources in Kubernetes, so that doesn't sway me.

**Comment by bgrant0607** [comment]
cc @kubernetes/sig-scheduling-api-reviews @kubernetes/sig-node-api-reviews

**Comment by k82cn** [comment]
Some questions, if already discussed, please ignore :)

1. How to define `unlimited`?
2. What's the behaviour if total/summary of ulimit is bigger than host?
3. For `IsUlimitResourceName`, can we add some prefix, e.g. `k8s.io`, to avoid conflict?


Ans also share the same comments with bgrant0607@ on whether ulimit should be a resource for container :)

**Comment by dims** [comment]
will revisit at some point ..

**Pull request closed without merging**